Sources/SwiftParser/Declarations.swift

```diff
@@ -1120,10 +1120,26 @@ extension Parser {
     _ handle: RecoveryConsumptionHandle
   ) -> RawDeinitializerDeclSyntax {
     let (unexpectedBeforeDeinitKeyword, deinitKeyword) = self.eat(handle)
+
     var unexpectedNameAndSignature: [RawSyntax?] = []
-    unexpectedNameAndSignature.append(self.consume(if: TokenSpec(.identifier, allowAtStartOfLine: false)).map(RawSyntax.init))
-    if self.at(.leftParen) && !self.currentToken.isAtStartOfLine {
-      unexpectedNameAndSignature.append(RawSyntax(parseFunctionSignature()))
+
+    // async is a contextual keyword
+    // must be parsed before attempting to parse identifier
+    var effectSpecifiers = parseDeinitEffectSpecifiers()
+
+    if effectSpecifiers == nil {
+      if let identifier = self.consume(if: TokenSpec(.identifier, allowAtStartOfLine: false)).map(RawSyntax.init) {
+        unexpectedNameAndSignature.append(identifier)
+      }
+      effectSpecifiers = parseDeinitEffectSpecifiers()
+    }
+    if effectSpecifiers == nil && self.at(.leftParen) && !self.currentToken.isAtStartOfLine {
+      let input = parseParameterClause(RawParameterClauseSyntax.self) { parser in
+        parser.parseFunctionParameter()
+      }
```

Comment:
If you’re changing this from `parseFunctionSignature` to `parseParameterClause`, I think we aren’t recovering the following as nicely anymore. Can’t we continue to use `parseFunctionSignature` here anymore? 

```swift
deinit() -> Int {}
```

And also: Could you add a test case for a `deinit` with a return value?

Reply:
`parseFunctionSignature()` uses `parseFunctionEffectSpecifiers()` which we don't want. I don't think it is worth to abstract `parseFunctionSignature()` over `parseFunctionEffectSpecifiers()` vs `parseDeinitEffectSpecifiers()`. It would be simpler to use `parseParameterClause()` + `parseDeinitEffectSpecifiers()` + `parseFunctionReturnClause()` directly from `parseDeinitializerDeclaration()`.

As a bonus it would allow recovering from `deinit -> Int {}` which `parseFunctionSignature()` cannot.

Note that using `RawDeinitEffectSpecifiersSyntax` with`parseFunctionReturnClause()` still requires conformance to  `RawEffectSpecifiersTrait` - that's still a bit ugly, but at least now without any logic.

Reply:
Updated, please take a look

Reply:
Oh, nice. That looks good to me.